kroxylicious/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/BrokerAddressPatternUtils.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal.clusternetworkaddressconfigprovider;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.regex.Pattern;
+
+class BrokerAddressPatternUtils {
+    public static final String LITERAL_NODE_ID = "$(nodeId)";
+    public static final Pattern LITERAL_NODE_ID_PATTERN = Pattern.compile(Pattern.quote(LITERAL_NODE_ID));
+    public static final Pattern PORT_SPECIFIER_RE = Pattern.compile(":([0-9]+)$");
+    public static final Pattern TOKEN_RE = Pattern.compile("(\\$\\([^)]+\\))");
+
+    static void validateTokens(String stringWithTokens,
```

Comment:
Think this might read better with `Set<String> tokens = extractTokens(stringWithTokens)` followed by the set logic.